packages/cli-kit/src/private/node/api.ts

```diff
@@ -30,30 +32,90 @@ function responseHeaderIsInteresting(header: string): boolean {
   return interestingResponseHeaders.has(header)
 }
 
-type VerboseResponse<T> = {
+interface CommonResponse {
   duration: number
   sanitizedHeaders: string
   sanitizedUrl: string
   requestId?: string
-} & (
-  | {status: 'ok'; response: T}
-  | {status: 'client-error'; clientError: ClientError}
-  | {status: 'unknown-error'; error: unknown}
-  | {status: 'can-retry'; clientError: ClientError; delayMs: number | undefined}
-  | {status: 'unauthorized'; clientError: ClientError; delayMs: number | undefined}
-)
+}
+
+type OkResponse<T> = CommonResponse & {status: 'ok'; response: T}
+type ClientErrorResponse = CommonResponse & {status: 'client-error'; clientError: ClientError}
+type UnknownErrorResponse = CommonResponse & {status: 'unknown-error'; error: unknown}
+type CanRetryErrorResponse = CommonResponse & {
+  status: 'can-retry'
+  clientError: ClientError
+  delayMs: number | undefined
+}
+type UnauthorizedErrorResponse = CommonResponse & {
+  status: 'unauthorized'
+  clientError: ClientError
+  delayMs: number | undefined
+}
+
+type VerboseResponse<T> =
+  | OkResponse<T>
+  | ClientErrorResponse
+  | UnknownErrorResponse
+  | CanRetryErrorResponse
+  | UnauthorizedErrorResponse
+
+function isARetryableNetworkError(error: unknown): boolean {
+  if (error instanceof Error) {
+    const networkErrorMessages = [
+      'socket hang up',
+      'ECONNRESET',
+      'ECONNABORTED',
+      'ENOTFOUND',
+      'ENETUNREACH',
+      'network socket disonnected',
+      'ETIMEDOUT',
+      'ECONNREFUSED',
+      'EAI_FAIL',
+    ]
+    const anyMatches = networkErrorMessages.some((issueMessage) => error.message.includes(issueMessage))
+    return anyMatches
+  }
+  return false
+}
+
+async function runRequestWithNetworkLevelRetry<T extends {headers: Headers; status: number}>({
+  request,
+  url,
+  enableNetworkLevelRetry,
+}: RequestOptions<T> & {enableNetworkLevelRetry: boolean}): Promise<T> {
+  if (!enableNetworkLevelRetry) {
+    return request()
+  }
+
+  let lastSeenError: unknown
+
+  for await (const _delayMs of sleepWithBackoffUntil()) {
+    try {
+      return await request()
+    } catch (err) {
+      lastSeenError = err
+      if (!isARetryableNetworkError(err)) {
+        throw err
+      }
+      outputDebug(`Retrying request to ${url} due to network error ${err}`)
+    }
+  }
+  throw lastSeenError
+}
 
 async function makeVerboseRequest<T extends {headers: Headers; status: number}>({
   request,
   url,
-}: RequestOptions<T>): Promise<VerboseResponse<T>> {
+  enableNetworkLevelRetry,
+}: RequestOptions<T> & {enableNetworkLevelRetry: boolean}): Promise<VerboseResponse<T>> {
   const t0 = performance.now()
   let duration = 0
   const responseHeaders: {[key: string]: string} = {}
   const sanitizedUrl = sanitizeURL(url)
   let response: T = {} as T
   try {
-    response = await request()
+    response = await runRequestWithNetworkLevelRetry({request, url, enableNetworkLevelRetry})
```

Comment:
Should we pass in `sanitizedUrl` here instead? It looks like our other `outputDebug` uses use it (e.g. `outputDebug('Request to ${result.sanitizedUrl} completed in ${result.duration} ms...`)

```suggestion
    response = await runRequestWithNetworkLevelRetry({request, sanitizedUrl, enableNetworkLevelRetry})
```